### DISCUSSION TITLE
Relay + (Status) LEDs

### DISCUSSION THREAD

**Discussion by andrewbpowers** [arendst/Tasmota#13426] [Q&A] [answered]
Hello!

I have only been working with Tasmota for a short time and the first small "project" is running wonderfully.
:-)

What I don't understand yet or have read somewhere clearly is a solution for the following "problem":

I would like to "build" a thermostat for Home Assistant. For this I use a relay, as well as a DHT22 and two LEDs. 
So far, the DHT22 is recognised and sends the data to Home Assistant via MQTT. I can also switch the relay on and off. 

But I would like LED No. 1, which is connected to GPIO0, to light up when the relay is "off" and switched off, when the relay is "on". LED No. 2, which is connected to GPIO1, should light up, when the relay is "on". When the relay is "off" again, LED No. 2 should go out and LED 1 should light up again.

Can I configure this via the UI or can I do it via the console or a config file?

Maybe the solution is quite simple, but at the moment I can't figure it out. :-(

Oh and one more question, when I select Led in the module parameters, for example, a new field appears to the right of it and I can select 1-4. What is this information for? 

Many thanks in advance!
:-)

**Comment by barbudor** [comment, marked as answer]
Tasmota can by default manage 1 LED associated to each Relay with the following condition:
- You need to define a GPIO for LedLink which is intended to be a Led which reports the Wifi/MQTT connection status. If you don't have a led for that or don't want such led, still you should assign LedLink to a free GPIO (unconnected) because it makes Tasmota behave as describe here after. If you don't, my explanations below won't work.
- Once the above condition is met, LED_1 will report the power status of Relay_1 automatically, same for LED_2 and Relay_2 etc ....
- If the on/off status of the led do not match what you want, you can use LED_i1 wher 'i' means inverted
- You can have only 1 LED_1, 1 LED_2 etc.... So it is not natively possible to let Tasmota manage automatically 2 LEDs for the same relay 
- However once you have the 1st led managed by Tasmota, you can use RULES (https://tasmota.github.io/docs/Rules/) to act on changing power states to drive the 2nd led
For example a rule-set made of the 2 following rules should makeit for driving LED_2:
```
on power1#state=0 do LedPower2 1 endon
on power1#state=1 do LedPower2 0 endon
```
You may have to invert 1 and 0 depending if you want the led on or off

Read carrefuly and multiple times the Rules chapter of the documentation as well as the description of the LedPower<x> command https://tasmota.github.io/docs/Commands/#ledpower

**Reply by andrewbpowers** [reply]
First of all, thank you very much for your answer and the solution. Really great! 
Big THANK you!

I use an ESP-01S and therefore all free GPIO are occupied.

What I don't quite understand is that when I switch the relay, I see ON or OFF in the console.

Why does your solution query the "LedLink" element and not the status of the relay?

I have assigned "LedLink" (LED 1) to GPIO0 and "Led" (LED 2) to GPIO1.
Relay is on GPIO2 and the DHT22 on 3.
Then I entered your two lines in the console as follows:
rule2 on power1#state=0 do LedPower2 1 endon
rule3 on power1#state=1 do LedPower2 0 endon

Unfortunately, that doesn't work... I can't switch the relay either.

**Reply by andrewbpowers** [reply]
With the following configuration, everything works as far as it can, except that LED2 is simply always on:

GPIO0: LED1 (Led_i) - 1
GPIO1: LED2 (Led) - 1
GPIO2: Relay - 1
GPIO3: DHT22

When I switch on the relay, LED1 goes off. When I switch it off, LED1 goes on again.
Unfortunately, LED2 is permanently lit.

**Reply by ascillato** [reply]
Are you using Tasmota version 10? If not, please update and re test. Thanks.

**Reply by andrewbpowers** [reply]
> Are you using Tasmota version 10? If not, please update and re test. Thanks.

I use Tasmota 10.0.0!

**Comment by barbudor** [comment]
You Must set a GPIO to LedLink as I explained, even one not usable
With ESP01S any GPIO that is not broken out to the edge connector is fine for that : 4, 5, 6, 7,... 12, 13, 14, 15, 16

**Reply by andrewbpowers** [reply]
Thanks again for your feedback!


I set D8 GPIO15 to LedLink...

GPIO0: LED1 (Led_i) - 1
GPIO1: LED2 (Led) - 1
GPIO2: Relay - 1
GPIO3: DHT22

...and the rules:
rule2 on power1#state=0 do LedPower2 1 endon
rule3 on power1#state=1 do LedPower2 0 endon

Nothing changes.

**Reply by barbudor** [reply]
> GPIO1: LED2 (Led) - 1

What does this means? Is it configured as LED_2?

As already said only 1 led for each index

**Reply by andrewbpowers** [reply]
Maybe I didn't describe exactly enough what I want or the "problem" is.

If the relay is "off", LED1 (blue, GPIO0) should be lit and LED2 (red, GPIO1) should be off.
If the relay is "on", LED1 (blue, GPIO0) should be off and LED2 (red, GPIO1) should be on.

I still don't understand the connection with LedLink.
I want to switch LEDs depending on the relay state and not the relay depending on whether any LED is on or off.

D3 GPIO0 - Led - 1 <- Index?
TX GPIO1 - Led - 2 <- Index?

So the field (with the 1-4 no.) appears to the right of the "Led" is the index of a Led and only the index is relevant for the rules?

Sorry, somehow I still don't understand how the connection is and I don't have found clear and simple examples of rules.

**Reply by barbudor** [reply]
Yes Led_1 means led component with index 1 in the 2nd dropdown

And you must have a GPIO assigned to LedLink, even unused gpio with nothing connected so you can use LedPower<x>

**Reply by andrewbpowers** [reply]
Ok!
Well, at least I understood something ;-)

It looks like the rules have an effect on the relay or the whole configuration, just like the assignment of the LED in the config module.
I can no longer switch the relay. 

I didn't think that such a simple problem would cause so much trouble. 

For the appropriate JSON file, for another firmware solution, I only needed two or three minutes and everything worked.

**Reply by andrewbpowers** [reply]
I reset the configuration... have started again and follow basic config made:

Module typ: Generic (18)
---
D3 GPIO0 - “Led_i” - “1”
TX GPIO1 - “Led” - “2”
D4 GPIO2 - “Relay” - “1”
RX GPIO3 - “SI7021”
…
D8 GPIO15 “LedLink”

After the restart, the blue LED (1) lights up, the red LED (2) is off and the relay is also off.
If I switch ON, the blue LED (1) switch off and the relay goes on. The red LED (2) is still off.
Ok, be so "good"! 

Unfortunately, no rule has worked so far, or at the latest after activating the rules, I could no longer switch the relay.

**Reply by barbudor** [reply]
Apologies I missed a little detail : as soon as you start to use the command `LedPower<x>` , Tasmota set LedState to 0 and decouple the leds from the relays. So Led_1 is no more attached to relay 1.
So you need the rule to control both LedPower1 and LedPower2.

Using different pins because I'm using a NodeMCU, but principle remais the same
- D5 GPIO14 Led_1 (not inverted because my led is connected between GPIO and GND. Use Led_i1 if your led is connected between GPIO and 3V3)
- D6 GPIO12  Led_2
- D7 GPIO13 Relay_1
- D8 GPIO15 LedLink

`rule1 on power1#state=0 do backlog ledpower1 1; ledpower2 0 endon on power1#state=1 do backlog ledpower1 0; ledpower2 1 endon`
It work as expected.
You can  swap Led_1/Led_2 or change which one use 0/1 or 1/0.
Sorry for the incomplete answer

**Reply by andrewbpowers** [reply]
Great! 
Thank you very much!
:-)

Your explanations were very good and helped me to first understand the basics, the correlations and to deal with "rules" at all.

I only changed [D3 GPIO0 - "Led_i" - "1"] --> [D3 GPIO0 - "Led" - "1"] in the configuration.
Then I entered the following rule in the console:
`rule1 on power1#state=0 do backlog ledpower1 1; ledpower2 0 endon on power1#state=1 do backlog ledpower1 0; ledpower2 1 endon`
...and activated with `Rule1 ON`.

Then a restart... and yeah, thanks to your great help, it now works the way I wanted it to.
:-)

**Comment by andrewbpowers** [comment]
Sorry, but I have another question.... 

Everything ran for a few days now without any problems and exactly as I wanted it to, with the settings described above. 

However, the following "phenomenon" has now crept in:
At some point, the ESP could no longer be addressed by Home Assistant.

After a restart (regardless of whether via the web interface or switching on and off), the ESP behaves as follows.
Relay is ON, status LEDs OFF. 

Only when I switch on the relay via the web interface does the red LED come on. 
When I then switch the relay off, the red LED goes off and the blue LED goes on. 
As it should be. Only then is the rule apparently applied. 

After each restart, the relay is always on and both LEDs are off.
However, I have not changed anything in the configuration.
Everything was running perfectly... even after a restart (relay off, blue LED on).

Apparently it doesn't matter what I set in the "PowerOnState". There I have set "Keep relay(s) off after power on".

What could be the error?

Many thanks for any help and tips!

**Reply by sfromis** [reply]
If you have been using retained commands, the broker will resend the last one after MQTT connect, even if you stopped setting retain on further commands. Logs at restart time (with `Weblog 3`) may have details about what happens why.

**Reply by andrewbpowers** [reply]
Hello!

Thanks for the tip about "Weblog 3"!
Almost everything is displayed with it. What do I have to watch out for?

**Reply by sfromis** [reply]
Watch what happens, try playing with on/off while watching the console, to get a feeling for it.

**Reply by andrewbpowers** [reply]
I don't know what's supposed to be "suspicious" in the log.

In addition, the error also occurs during restart and not during operation after I have clicked ON and then OFF.
From then on, everything is normal again, as described.

**Reply by sfromis** [reply]
Just watch what happens..... If you catch it right away, it may still be in the console. Or you can activate remote logging with SysLog or MqttLog commands. Use log level 3 for extra details.

**Reply by andrewbpowers** [reply]
Thank you for your feedback!

I can't imagine right now that it's MQTT that's responsible for the state after the restart.
If I completely disconnect the ESP from the power supply and then switch it on again, I have exactly the same problem.
At that point, there is still no MQTT connection.

**Reply by ascillato** [reply]
What you see is just a very fast blink in the same moment you power up your device?

If so, that is just that you are using the GPIOs that flips at boot time. That is just hardware and can not be controlled by the software. The solution is to use a different GPIO. See Tasmota docs for more information of GPIOs at Boot time in ESP8266

**Reply by andrewbpowers** [reply]
Thank you for your information!

Everything ran for days without any problems and as it should... even after several reboots.
The relay is switched on exactly at the moment when I click on reboot. And then it boots normally.

**Reply by ascillato** [reply]
It changes its state and remain in that state or it just blinks? If it is a blink, it can't be solved by software

**Reply by andrewbpowers** [reply]
Nothing blinks. And once again, everything worked for several days.
Even with a restart or when the ESP was switched off completely. So there must be a reason why this error occurs at some point.

And one more thing... I flashed another ESP, loaded the saved config... and... everything works. The only question is, for how long?

**Reply by ascillato** [reply]
So, you have to check the logs. Something is triggering that. Please, set weblog to 4, then restart your device and paste here the log.

**Reply by andrewbpowers** [reply]
Thank you, for your feedback!

A small update...
I have exactly the same problem with another ESP-01S after some time. I have absolutely no idea what the problem is.

Now I have installed a D1 Mini and found the following...
PowerOnState really only works when you switch off the system completely, but not during a reset.
The rule used above no longer worked. Apparently the rule is not applied immediately after a restart or reset.

The pins are now connected as follows:
- TX GPIO1 - Led_i - 1
- D4 GPIO - SI7021
- RX GPIO3 Led - 2
- D1 GPIO - Relay_i - 1
- D0 GPIO16 - LedLink

The applied rule is now:
`rule1 on power1#state=0 do backlog ledpower1 0; ledpower2 0 endon on power1#state=1 do backlog ledpower1 1; ledpower2 1 endon`
